File: leetcode/p00706/p00706.py
class Node:
    def __init__(self, key=0, val=0):
        self.key = key
        self.val = val
class MyHashMap:
    def __init__(self):
        self.buckets = [[] for _ in range(1 << 13)]

    def hasher(self, key: int) -> int:
        return ((key*786433) & (1<<20-1)) >> 7

    def put(self, key: int, value: int) -> None:
        hashed = self.hasher(key)
        
        for i in range(len(self.buckets[hashed])):
            if self.buckets[hashed][i].key == key:
                self.buckets[hashed][i].val = value
                return

        self.buckets[hashed].append(Node(key, value))

    def get(self, key: int) -> int:
        hashed = self.hasher(key)
        for i in range(len(self.buckets[hashed])):
            if self.buckets[hashed][i].key == key:
                return self.buckets[hashed][i].val

        return -1

    def remove(self, key: int) -> None:
        hashed = self.hasher(key)
        
        idx = -1
        for i in range(len(self.buckets[hashed])):
            if self.buckets[hashed][i].key == key:
                idx = i
                break
        
        if idx == -1:
            return
        if idx == 0:
            self.buckets[hashed] = self.buckets[hashed][1:]
        elif idx == len(self.buckets[hashed])-1:
            self.buckets[hashed].pop()
        else:
            self.buckets[hashed] = self.buckets[hashed][:idx] + self.buckets[hashed][idx+1:]

if __name__ == '__main__':
    obj = MyHashMap()
    
    obj.put(1,1)
    obj.put(2,2)
    assert obj.get(1) == 1
    assert obj.get(3) == -1
    obj.put(2,1)
    assert obj.get(2) == 1
    obj.remove(2)
    assert obj.get(2) == -1

File: leetcode/p00706/p00706.go
package p00706

type Node struct {
	Key int
	Val int
}
type MyHashMap struct {
	buckets [][]*Node
}

func hasher(key int) int {
	return ((key * 786433) & (1<<20 - 1)) >> 7
}

func Constructor() MyHashMap {
	return MyHashMap{
		buckets: make([][]*Node, 1<<13),
	}
}

func (this *MyHashMap) Put(key int, value int) {
	hashed := hasher(key)

	for _, node := range this.buckets[hashed] {
		if node.Key == key {
			node.Val = value
			return
		}
	}
	this.buckets[hashed] = append(this.buckets[hashed], &Node{Key: key, Val: value})
}

func (this *MyHashMap) Get(key int) int {
	for _, node := range this.buckets[hasher(key)] {
		if node.Key == key {
			return node.Val
		}
	}

	return -1
}

func (this *MyHashMap) Remove(key int) {
	hashed := hasher(key)

	idx := -1
	for i, node := range this.buckets[hashed] {
		if node.Key == key {
			idx = i
			break
		}
	}

	if idx == -1 {
		return
	} else if idx == 0 {
		this.buckets[hashed] = append([]*Node{}, this.buckets[hashed][1:]...)
	} else if idx == len(this.buckets[hashed])-1 {
		this.buckets[hashed] = append([]*Node{}, this.buckets[hashed][:idx]...)
	} else {
		this.buckets[hashed] = append(this.buckets[hashed][:idx], this.buckets[hashed][idx+1:]...)
	}
}
